### PR TITLE
Correct the two-stage parsing strategy of antlr parser

### PR DESCRIPTION
**Pull request by pan3793**
## Description

This PR follows the https://github.com/antlr/antlr4/issues/192#issuecomment-15238595 to correct the current implementation of the **two-stage parsing strategy** in `SqlParser`.

> You can save a great deal of time on correct inputs by using a two-stage parsing strategy.
>
> - Attempt to parse the input using BailErrorStrategy and PredictionMode.SLL.
>    If no exception is thrown, you know the answer is correct.
> - If a ParseCancellationException is thrown, retry the parse using the default
>    settings (DefaultErrorStrategy and PredictionMode.LL).

## Additional context and related issues

We recently identified and fixed this issue in Apache Spark https://github.com/apache/spark/pull/40835

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(x) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

**Comment by pan3793** [comment]
CLA has been signed and sent via email.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParserErrorHandling.java [hunk 1]
```diff
@@ -126,7 +126,7 @@ private static Stream<Arguments> statements()
                         "line 1:19: mismatched input 'x'. Expecting: '(', ',', 'CROSS', 'EXCEPT', 'FETCH', 'FULL', 'GROUP', 'HAVING', 'INNER', 'INTERSECT', 'JOIN', 'LEFT', 'LIMIT', " +
                                 "'MATCH_RECOGNIZE', 'NATURAL', 'OFFSET', 'ORDER', 'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', 'WINDOW', <EOF>"),
                 Arguments.of("SELECT * FROM t WHERE EXISTS (",
-                        "line 1:31: mismatched input '<EOF>'. Expecting: <query>"),
+                        "line 1:31: mismatched input '<EOF>'. Expecting: '(', 'SELECT', 'TABLE', 'VALUES'"),
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParserErrorHandling.java:129]: We want to retain `<query>` in this case.

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java [hunk 2]
```diff
@@ -264,7 +264,7 @@ public void testRenameColumnName(String columnName)
         }
         assertThatThrownBy(() -> super.testRenameColumnName(columnName))
                 // TODO (https://github.com/trinodb/trino/issues/7205) support column rename in Phoenix
-                .hasMessageContaining("Syntax error. Encountered \"RENAME\"");
+                .hasMessageContaining("ERROR 722 (43M05): Schema does not exist schemaName=TPCH");
```
Comment (martint) [inline plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:267]: This seems unrelated.

**Comment by mosabua** [comment]
Do we want to continue with this @martint ?
